### PR TITLE
Add LLVM codegen support for integer matmuls

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
This adds generic lowering for linalg tensor matmuls based on the
contraction op interface. For the buffers path, we just hardcode a few
ops, since hopefully that's just getting deleted soon.

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
This depends on some recent MLIR and MHLO stuff, so currently based on the google branch with an additional submodule update. That will be dropped before merging once `main` is updated, so only 40752be is really part of this PR.

File: iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToLLVM/KernelDispatch.cpp:73]: THank you for cleaning up this eye sore :) !

**Review by MaheshRavishankar — APPROVED** [review]
Nice! LGTM and awesome cleanup. I have been eyeing those template specializations for a while!
Maybe wait for Ahmed to LGTM too cause those were added to allow for getting the values from command lines.

**Review by asaadaldien — APPROVED** [review]
AWESOME!

**Comment by GMNGeoffrey** [comment]
Whew, I was a bit worried that I was deleting something important

**Comment by asaadaldien** [comment]
> Whew, I was a bit worried that I was deleting something important

Its not important and serving the same propose. In practice ops we will specified here will have contraction interface (if this part survived any longer)

**Pull request merged**
